components/kubeflow-pipeline/python/fl_suite/pipelines/_pipelines.py

```diff
@@ -32,12 +32,12 @@ def _pipeline(
     # pylint: disable=too-many-arguments
     @dsl.pipeline()
     def fl_pipeline(
-        image_tag: str,
-        num_rounds: int,
-        num_local_rounds: int,
-        min_available_clients: int,
-        min_fit_clients: int,
-        min_eval_clients: int,
+        image_tag: str = _create_image_tag("fl-client"),
```

Comment:
This doesn't work as you would expect: the `_create_image_tag("fl-client")` is called only once when the function is defined. The next times will re-use that first value from `_create_image_tag("fl-client")`. Instead do:

```python
@dsl.pipeline()
def fl_pipeline(
    image_tag: Optional[str] = None
    # ...
):
    if image_tag is None:
        image_tag = _create_image_tag("fl-client")
    # ...
```

Reply:
That approach won't work with `pipelines.build` though, as the default values are extracted from the `dsl.pipeline` decorator. As the `fl_pipeline` function is defined when `_pipeline` is called, wouldn't that result in what we want?

Reply:
Then I got it wrong. The idea is to `_create_image_tag` only once, so that the `fl_pipeline` uses the same image always, e.g

```python
# It is expected that these 2 calls:
fl_pipeline()
fl_pipeline()

# Do this
fl_pipeline(
  image_tag = "fl_client:2022.06.02-10.30" # <--- notice that this will not change once is defined
  num_rounds: int = 3,
  num_local_rounds: int = 1,
  min_available_clients: int = 2,
  min_fit_clients: int = 2,
  min_eval_clients: int = 2,
)
```

If that's the case, then you are right.

Reply:
Yes, the expectation is that call `pipelines.build` will create a pipeline YAML file that has a good default value for the pipeline. `pipelines.run` won't be affected here, as there the image tag is overwritten every time the function is called.

Reply:
Great. Then the PR is good to go. Thanks!